File: drills/elixir/solutions/nc115_redundant_connection__by_removal.exs
defmodule Solution do
  # Try removing each edge, latest first, and keep the first removal that leaves
  # a tree. O(n^2) against union-find's near-linear, but it needs no new
  # structure -- and it says the specification outright: the answer is the last
  # edge whose absence would make the graph a tree.
  def find_redundant_connection(edges) do
    nodes = edges |> List.flatten() |> MapSet.new()
    indexed = Enum.with_index(edges)

    indexed
    |> Enum.reverse()
    |> Enum.find_value([], fn {edge, i} ->
      remaining = for {other, j} <- indexed, j != i, do: other
      if tree?(remaining, nodes), do: edge, else: nil
    end)
  end

  defp tree?(edges, nodes) do
    cond do
      MapSet.size(nodes) == 0 ->
        true

      length(edges) != MapSet.size(nodes) - 1 ->
        false

      true ->
        adjacency =
          Enum.reduce(edges, %{}, fn [a, b], acc ->
            acc |> Map.update(a, [b], &[b | &1]) |> Map.update(b, [a], &[a | &1])
          end)

        seen = walk([Enum.min(nodes)], adjacency, MapSet.new())
        MapSet.size(seen) == MapSet.size(nodes)
    end
  end

  defp walk([], _adjacency, seen), do: seen

  defp walk([node | rest], adjacency, seen) do
    if MapSet.member?(seen, node) do
      walk(rest, adjacency, seen)
    else
      walk(Map.get(adjacency, node, []) ++ rest, adjacency, MapSet.put(seen, node))
    end
  end
end

File: drills/go/solutions/nc115_redundant_connection__by_removal.go
package main

func findRedundantConnection(edges [][]int) []int {
	// Try removing each edge from the last to the first; the first removal
	// that leaves the graph a tree (connected, n-1 edges) is the answer.
	n := len(edges)
	for skip := n - 1; skip >= 0; skip-- {
		adjacent := make([][]int, n+1)
		for i, edge := range edges {
			if i == skip {
				continue
			}
			adjacent[edge[0]] = append(adjacent[edge[0]], edge[1])
			adjacent[edge[1]] = append(adjacent[edge[1]], edge[0])
		}
		seen := make([]bool, n+1)
		stack := []int{1}
		seen[1] = true
		reached := 0
		for len(stack) > 0 {
			node := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			reached++
			for _, next := range adjacent[node] {
				if !seen[next] {
					seen[next] = true
					stack = append(stack, next)
				}
			}
		}
		if reached == n {
			return edges[skip]
		}
	}
	return []int{}
}
